sw/device/tests/crypto/cryptotest/firmware/otbn_fi.c

```diff
@@ -0,0 +1,261 @@
+  otbn_fi_loop_counter_t uj_output;
+  uj_output.loop_counter = loop_counter;
+  uj_output.err_status = err_bits;
+  RESP_OK(ujson_serialize_otbn_fi_loop_counter_t, uj, &uj_output);
+  return OK_STATUS(0);
+}
+
+/**
+ * otbn.fi.char.unrolled.dmem.op.loop command handler.
+ *
+ * This FI penetration tests executes the following instructions on OTBN:
+ * - Perform 100 times:
+ *  - Load loop counter from memory
+ *  - Increment loop counter
+ *  - Store back to memory
+ * - Return the value over UART.
+ *
+ * Faults are injected during the trigger_high & trigger_low.
+ * It needs to be ensured that the compiler does not optimize this code.
+ *
+ * @param uj The received uJSON data.
+ */
+status_t handle_otbn_fi_char_unrolled_dmem_op_loop(ujson_t *uj) {
+  // Initialize OTBN app, load it, and get interface to OTBN data memory.
+  OTBN_DECLARE_APP_SYMBOLS(otbn_char_unrolled_dmem_op_loop);
+  OTBN_DECLARE_SYMBOL_ADDR(otbn_char_unrolled_dmem_op_loop, lc);
+  const otbn_app_t kOtbnAppCharUnrolledDmemOpLoop =
+      OTBN_APP_T_INIT(otbn_char_unrolled_dmem_op_loop);
+  static const otbn_addr_t kOtbnAppCharUnrolledDmemOpLoopLC =
+      OTBN_ADDR_T_INIT(otbn_char_unrolled_dmem_op_loop, lc);
+  otbn_load_app(kOtbnAppCharUnrolledDmemOpLoop);
+
+  uint32_t loop_counter;
+
+  // FI code target.
+  sca_set_trigger_high();
+  otbn_execute();
+  otbn_busy_wait_for_done();
+  sca_set_trigger_low();
+
+  // Read loop counter from OTBN data memory.
+  otbn_dmem_read(1, kOtbnAppCharUnrolledDmemOpLoopLC, &loop_counter);
+
+  // Read ERR_STATUS register from OTBN.
+  dif_otbn_err_bits_t err_bits;
+  read_otbn_err_bits(&err_bits);
+
+  // Send loop counter & ERR_STATUS to host.
+  otbn_fi_loop_counter_t uj_output;
+  uj_output.loop_counter = loop_counter;
+  uj_output.err_status = err_bits;
+  RESP_OK(ujson_serialize_otbn_fi_loop_counter_t, uj, &uj_output);
+  return OK_STATUS(0);
+}
+
+/**
+ * otbn.char.unrolled.reg.op.loop command handler.
+ *
+ * This FI penetration tests executes the following instructions on OTBN:
+ * - Initialize register x2=0
+ * - Perform 100 x2 = x2 + 1 additions
+ * - Return the value over UART.
+ *
+ * Faults are injected during the trigger_high & trigger_low.
+ * It needs to be ensured that the compiler does not optimize this code.
+ *
+ * @param uj The received uJSON data.
+ */
+status_t handle_otbn_fi_char_unrolled_reg_op_loop(ujson_t *uj) {
+  // Initialize OTBN app, load it, and get interface to OTBN data memory.
+  OTBN_DECLARE_APP_SYMBOLS(otbn_char_unrolled_reg_op_loop);
+  OTBN_DECLARE_SYMBOL_ADDR(otbn_char_unrolled_reg_op_loop, lc);
+  const otbn_app_t kOtbnAppCharUnrolledRegOpLoop =
+      OTBN_APP_T_INIT(otbn_char_unrolled_reg_op_loop);
+  static const otbn_addr_t kOtbnAppCharUnrolledRegOpLoopLC =
+      OTBN_ADDR_T_INIT(otbn_char_unrolled_reg_op_loop, lc);
+  otbn_load_app(kOtbnAppCharUnrolledRegOpLoop);
+
+  uint32_t loop_counter;
+
+  // FI code target.
+  sca_set_trigger_high();
+  otbn_execute();
+  otbn_busy_wait_for_done();
```

Comment:
Thanks for the review! Yes, this function should also end when FI did some harm inside OTBN. In that case, the `otbn_busy_wait_for_done()` function returns an error code and does not affect the surrounding code. I am not storing this error code because I am already reading the error register below using `read_otbn_err_bits()` - this should give me more information what went wrong.

Reply:
Great, shall I merge this?

Reply:
Yes, please :-)